File: Medium/152_Maximum_Product_Subarray/maximum_product_subarray.py
from typing import List
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        min_ending = nums[0]
        max_ending = nums[0]
        maximum_product =  max_ending

        for num in nums[1:]:
            temp = min_ending
            min_ending = min(num, num*min_ending, num*max_ending)
            max_ending = max(num, num*temp, num*max_ending)
            maximum_product = max(max_ending, maximum_product)
        return maximum_product

File: Medium/152_Maximum_Product_Subarray/maximum_product_subarray.go
package maximumproductsubarray

func maxProduct(nums []int) int {
	curMaxProduct := nums[0]
	curMinProduct := nums[0]
	maxSoFar := nums[0]

	for i := 1; i < len(nums); i++ {
		temp := curMaxProduct
		curMaxProduct = max(nums[i], curMaxProduct*nums[i], curMinProduct*nums[i])
		curMinProduct = min(nums[i], temp*nums[i], curMinProduct*nums[i])
		maxSoFar = max(maxSoFar, curMaxProduct, curMinProduct)
	}

	return maxSoFar
}
